lib/Parse/ParseDecl.cpp

```diff
@@ -770,31 +770,72 @@ Parser::parseDifferentiableAttribute(SourceLoc atLoc, SourceLoc loc) {
   ParserStatus Status;
   SourceLoc lParenLoc, rParenLoc;
 
-  // Set parse error, skip until ')' and parse it.
-  auto errorAndSkipToEnd = [&](int parenDepth = 1) -> ParserStatus {
+  // Parse '('.
+  if (!consumeIf(tok::l_paren, lParenLoc)) {
+    diagnose(Tok, diag::attr_expected_lparen, AttrName,
+             /*DeclModifier=*/false);
+    skipUntil(tok::r_paren);
+    if (!consumeIf(tok::r_paren, rParenLoc))
+      diagnose(Tok, diag::attr_expected_rparen, AttrName,
+               /*DeclModifier=*/false);
+    return makeParserError();
+  }
+
+  using FuncSpec = DifferentiableAttr::FunctionSpecifier;
+  AutoDiffMode mode;
+  SourceLoc modeLoc;
+  SmallVector<AutoDiffParameter, 8> params;
+  Optional<FuncSpec> primalSpec;
+  FuncSpec adjointSpec;
+  TrailingWhereClause *whereClause = nullptr;
+
+  // Parse @differentiable attribute arguments.
+  if (parseDifferentiableAttributeArguments(mode, modeLoc, params, primalSpec,
+                                            adjointSpec, whereClause))
+    return makeParserError();
+
+  // Parse ')'.
+  if (!consumeIf(tok::r_paren, rParenLoc)) {
+    diagnose(getEndOfPreviousLoc(), diag::attr_expected_rparen, AttrName,
+             /*DeclModifier=*/false);
     Status.setIsParseError();
+    return Status;
```

Comment:
Is there a difference between `return makeParserError()` and `Status.setIsParserError(); return Status;`? And is there a reason that some places in this function do one and others do the other?

Reply:
AFAIK, there's no difference. Changed to `return makeParserError()`.